infer/resource.go

```diff
@@ -865,41 +865,50 @@ func (*derivedResourceController[R, I, O]) getInstance() *R {
 }
 
 func (rc *derivedResourceController[R, I, O]) Check(ctx context.Context, req p.CheckRequest) (p.CheckResponse, error) {
-	encoder, i, failures, err := decodeCheckingMapErrors[I](req.News)
-	if err != nil {
-		return p.CheckResponse{}, err
-	}
-	if len(failures) > 0 {
-		return p.CheckResponse{
-			// If we failed to decode, we apply secrets pro-actively to ensure
-			// that they don't leak into previews.
-			Inputs:   applySecrets[I](req.News),
-			Failures: failures,
-		}, nil
-	}
-
 	var r R
 	if r, ok := ((interface{})(r)).(CustomCheck[I]); ok {
 		// The user implemented check manually, so call that.
 		//
 		// We do not apply defaults or secrets if the user has implemented Check
 		// themselves. Defaults and secrets are applied by [DefaultCheck].
 
-		defCheckEnc, i, failures, err := callCustomCheck(ctx, r, req.Urn.Name(), req.Olds, req.News)
+		backupEncoder, _, _, _ := decodeCheckingMapErrors[I](req.News)
```

Comment:
Are you sure `failures`, `err` need no handling here? 

What if the backupEncoder moves under this `if` block:

```
if encoder == nil {
			encoder = &backupEncoder
}
```

So as to compute it lazily if really needed, and it that case still may handle the error?

Reply:
+1 - it may be correct now but doesn't look robust

Reply:
> Are you sure failures, err need no handling here?

I'll explain why in a comment. The tldr is: the backup encoder might not be derived from a shape that matches the shape of `I`, so it might not fit correctly. It will default to just doing a reasonable thing.